core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -358,12 +377,45 @@ final class ShuffleBlockFetcherIterator(
     }
   }
 
+
+  /**
+   * Fetch the host-local blocks while we are fetching remote blocks. This is ok because
+   * `ManagedBuffer`'s memory is allocated lazily when we create the input stream, so all we
+   * track in-memory are the ManagedBuffer references themselves.
+   */
+  private[this] def fetchHostLocalBlocks() {
+    logDebug(s"Start fetching host-local blocks: ${hostLocalBlocks.mkString(", ")}")
+
+    val localDirsByExec = blockManager.getHostLocalDirs(hostLocalBlocksByExecutor.keySet.toArray)
+    for (
+      (execId, blocks) <- hostLocalBlocksByExecutor;
+      localDirs = localDirsByExec(execId);
+      blockId <- blocks
```

Comment:
probably stating the obvious, but when you do change it so that the driver does not store an unbounded set of local dirs, then you'll have to do some error handling here to make sure you go back to doing a remote read.  worth a test case too